org.bndtools.launch/src/bndtools/launch/AbstractOSGiLaunchDelegate.java

```diff
@@ -171,70 +182,153 @@ public boolean buildForLaunch(ILaunchConfiguration configuration, String mode, I
+			.getLaunches()) {
+
+			if (configuration.equals(l.getLaunchConfiguration())) {
+				if (launchResource.equals(LaunchUtils.getTargetResource(l.getLaunchConfiguration()))) {
 					for (IProcess process : l.getProcesses()) {
-						if (!process.isTerminated())
-							processCount++;
+						if (!process.isTerminated()) {
+							return true;
+						}
 					}
 				}
 			}
+		}
+		return false;
+	}
 
-			// Warn if existing processes running
-			if (processCount > 0) {
-				Status status = new Status(IStatus.WARNING, PLUGIN_ID, 0,
-					"One or more OSGi Frameworks have already been launched for this configuration. Additional framework instances may interfere with each other due to the shared storage directory.",
-					null);
-				IStatusHandler prompter = DebugPlugin.getDefault()
-					.getStatusHandler(status);
-				if (prompter != null) {
-					boolean okay = (Boolean) prompter.handleStatus(status, launchResource);
-					if (!okay)
-						return okay;
-				}
-			}
+	private boolean checkProjectErrors(ILaunchConfiguration configuration, String mode, IProgressMonitor monitor)
+		throws CoreException {
+		monitor.beginTask("", 1); //$NON-NLS-1$
+		try {
+			return Central.getWorkspace()
+				.readLocked(() -> {
+					IProject[] projects = getRelatedProjects();
+
+					monitor.subTask(DebugCoreMessages.LaunchConfigurationDelegate_6);
+					List<IAdaptable> errors = new ArrayList<>();
+					for (IProject project : projects) {
+						monitor.subTask(
+							MessageFormat.format(DebugCoreMessages.LaunchConfigurationDelegate_7, new Object[] {
+								project.getName()
+						}));
+						if (existsProblems(project)) {
+							errors.add(project);
+						}
+					}
+					if (!errors.isEmpty()) {
+						errors.add(0, configuration);
+						IStatusHandler prompter = DebugPlugin.getDefault()
+							.getStatusHandler(promptStatus);
+						if (prompter != null) {
+							return ((Boolean) prompter.handleStatus(complileErrorProjectPromptStatus, errors))
+								.booleanValue();
+						}
+					}
+
+					return true;
+				});
+		} catch (Exception e) {
+			throw new CoreException(Status.error("Unexpected while checking compile errors", e));
+		} finally {
+			monitor.done();
 		}
 
-		IStatus launchStatus = getLauncherStatus();
+	}
 
-		IStatusHandler prompter = DebugPlugin.getDefault()
-			.getStatusHandler(launchStatus);
-		if (prompter != null)
-			return (Boolean) prompter.handleStatus(launchStatus, run);
-		return true;
+	private IProject[] getRelatedProjects() throws CoreException {
+		ProjectLauncher projectLauncher = getProjectLauncher();
+		if (projectLauncher == null)
+			return new IProject[0];
+
+		Project p = projectLauncher.getProject();
+		if (p == null)
+			return new IProject[0];
+
+		if (p instanceof Run) {
+			return getAllIProjects(p.getWorkspace());
+		}
+		return Central.getProject(p)
+			.map(x -> new IProject[] {
+				x
+			})
+			.orElse(new IProject[0]);
+	}
+
+	private IProject[] getAllIProjects(Workspace ws) {
```

Comment:
Any reason to not make this `static`? Seems to be a helper method that doesn't access any non-static fields.